### DISCUSSION TITLE
Collections, relations, typed properties and constructor assignment

### DISCUSSION THREAD

**Discussion by pkly** [doctrine/doctrine2#8833] [Support Questions]
Hi, I have a question about collections.
Assuming a ManyToMany or OneToMany relation between entities, one would create something like the following

```php
/**
 * @var Collection<int, Entity>|Entity[]
 * @ORM\OneToMany(targetEntity=Entity::class, mappedBy="myField")
 */
private $relationList;

public function __construct() {
    $this->relationList = new ArrayCollection();
}

public function getRelationList(): Collection
{
    return $this->relationList;
}
```

And this works fine, assuming _you_ create the entity.
Something weird happens however if you assume the following scenario.

```php
/**
 * @var Collection<int, Entity>|Entity[]
 * @ORM\OneToMany(targetEntity=Entity::class, mappedBy="myField")
 */
private Collection $relationList;

public function __construct() {
    $this->relationList = new ArrayCollection();
}

public function getRelationList(): Collection
{
    return $this->relationList;
}

// other file

$object = $repository->find(1);
$object->getRelationList(); // crash
```

In some scenarios accessing a field which is a multiple relation will try to access the field itself, which is not initialized and causes a crash. Why is that? Why do doctrine devs recommend creating collections in the constructor if it's not even fired during load from database? How should this be handled properly? Via the PostLoad hook and creating an empty collection?
Using doctrine/orm 2.9.3

**Comment by greg0ire** [comment]
Please make sure you use an up-to-date version of `doctrine/persistence`.

**Reply by pkly** [reply]
I just checked and we're on 2.2.1 so.
Afaik this is a design decision, as doctrine does not call a constructor, so the default creation of collections does not occur. As I understand this is because the collections that exist in the class will be replaced by and instance of the database collection anyway, so fair enough, but I've seen this error happen multiple times, so obviously there must be something amiss.

Do you maybe know if what I described here is even allowed behavior?
If yes, I'll attempt to reproduce it.

**Reply by greg0ire** [reply]
When I read this, I think it's supposed to be supported:

https://github.com/doctrine/persistence/blob/2272edb2ac3cdd400f972f0d7fc27a775b0af693/lib/Doctrine/Persistence/Reflection/TypedNoDefaultReflectionPropertyBase.php#L18-L20

Maybe it works for simple properties but not for collections somehow? IMO this is more like a bug and you should file one to `doctrine/persistence`, this time with a stack trace of the crash.

**Comment by beberlei** [comment]
This should work, it sounds like a bug. The `UnitOfWork` instantiates all collection properties that are mapped as such by setting them as `PersistentCollcetion` instance. You could Xdebug into `UnitOfWork::createEntity` and look whats happening around assocation mappings there. It might be related to the typed property, can you not make it typed and test if the behavior changes?

**Reply by pkly** [reply]
I'll need to get around to actually testing it. I know for a fact it can be caused somehow, but it's rather annoying to reproduce. Our installation script broke when using typed properties for collections, that I know for sure.
